README.md
# tailwindcss-shadcn-ui

<p align="center">
  <a href="https://www.npmjs.com/package/tailwindcss-shadcn-ui">
    <img alt="NPM Version" src="https://badgen.net/npm/v/tailwindcss-shadcn-ui" />
  </a>
  <a href="https://www.npmjs.com/package/tailwindcss-shadcn-ui">
    <img alt="Types Included" src="https://badgen.net/npm/types/tailwindcss-shadcn-ui" />
  </a>
  <a href="https://github.com/tobiasfoerg/tailwindcss-shadcn-ui/blob/main/LICENSE">
    <img alt="Apache-2.0 License" src="https://badgen.net/github/license/tobiasfoerg/tailwindcss-shadcn-ui" />
  </a>
  <a href="https://www.npmjs.com/package/tailwindcss-shadcn-ui">
    <img alt="NPM Downloads" src="https://badgen.net/npm/dm/tailwindcss-shadcn-ui" />
  </a>
</p>
<hr/>

A plugin/preset for Tailwind CSS v3.2+ that provides base utilities for shadcn ui.

## Installation

This plugin requires `tailwindcss` and `tailwindcss-animate`

**npm**

```sh
npm install tailwindcss-shadcn-ui --save-dev
```

**yarn**

```sh
yarn install -D tailwindcss-shadcn-ui
```

**pnpm**

```sh
pnpm add -D tailwindcss-shadcn-ui
```

## Configuration

### Recommend

Add the shadcn preset to your tailwind config file unsing the preset creator function.

> With this approach we can get a full configured preset including `tailwindcss-animate` plugin and dark mode set to class. Shoutout to [@simonswiss](https://twitter.com/simonswiss). He metioned the preset way in his video.

```ts
// tailwind.config.ts
import { type Config } from "tailwindcss";
import { createPreset } from "tailwindcss-shadcn-ui";

export default {
    presets: [createPreset()],
    // ...
} satisfies Config;
```

### Alternative

With this approach we can not preconfigure the plugins to include the `tailwindcss-animate` plugin. So you have to include it yourself.

<details>
  <summary>Using the plugin.</summary>

```ts
// tailwind.config.ts
import { type Config } from "tailwindcss";
import tailwindAnimate from "tailwindcss-animate";
import tailwindShadcnUi from "tailwindcss-shadcn-ui";

export default {
    darkMode: ["class"],
    // ...
    plugins: [
        // ...
        tailwindShadcnUi,
        tailwindAnimate,
    ],
} satisfies Config;
```

</details>

## Theme presets

```typescript
// tailwind.config.ts
import { createPreset } from "tailwindcss-shadcn-ui";
import { DEFAULT } from "tailwindcss-shadcn-ui/themes";

export default {
    presets: [createPreset({ theme: DEFAULT })],
    // ..
};
```

<details>
  <summary>Plugin and theme.</summary>

```typescript
// tailwind.config.ts
import tailwindShadcnUi from "tailwindcss-shadcn-ui";
import { DEFAULT } from "tailwindcss-shadcn-ui/themes";

export default {
    // ...
    plugins: [
        // ...
        tailwindShadcnUi({
            theme: DEFAULT,
        }),
    ],
};
```

</details>

### Available presets

| Preset  | Files                                        |
| ------- | -------------------------------------------- |
| DEFAULT | [themes/default.ts](./src/themes/default.ts) |

### Define your own theme

```typescript
// tailwind.config.ts

import { createPreset, defineTheme } from "tailwindcss-shadcn-ui";
import { DEFAULT } from "tailwindcss-shadcn-ui/themes";

export default {
    presets: [
        createPreset({
            theme: defineTheme({
                base: {
                    radius: "2rem", // string|number(pixels),
                    fontFamily: "Inter", // string|string[]
                },
                light: {
                    background: "0 0% 0%", // `${number} ${number}% ${number}%`
                    foreground: [0, 0, 0], // [hue: number, saturation: number, lightness: number]
                    //...
                },
                dark: {
                    background: "#000000", // `#${string}`
                    // ...
                },
            }),
        }),
    ],
    // ..
};

export default {
    // ...
    plugins: [
        // ...
        tailwindShadcnUi({
            theme: defineTheme({
                // ..
            }),
        }),
    ],
};
```

### The result

the preset adds following tailwind config.

```typescript
export default {
    plugins: [tailwindAnimate],
    theme: {
        container: {
            center: true,
            padding: "2rem",
            screens: {
                "2xl": "1400px",
            },
        },
        extend: {
            fontFamily: {
                sans: ["var(--font-sans)", ...fontFamily.sans],
            },
            colors: {
                border: "hsl(var(--border) / <alpha-value>)",
                input: "hsl(var(--input) / <alpha-value>)",
                ring: "hsl(var(--ring) / <alpha-value>)",
                background: "hsl(var(--background) / <alpha-value>)",
                foreground: "hsl(var(--foreground) / <alpha-value>)",
                primary: {
                    DEFAULT: "hsl(var(--primary) / <alpha-value>)",
                    foreground: "hsl(var(--primary-foreground) / <alpha-value>)",
                },
                secondary: {
                    DEFAULT: "hsl(var(--secondary) / <alpha-value>)",
                    foreground: "hsl(var(--secondary-foreground) / <alpha-value>)",
                },
                destructive: {
                    DEFAULT: "hsl(var(--destructive) / <alpha-value>)",
                    foreground: "hsl(var(--destructive-foreground) / <alpha-value>)",
                },
                muted: {
                    DEFAULT: "hsl(var(--muted) / <alpha-value>)",
                    foreground: "hsl(var(--muted-foreground) / <alpha-value>)",
                },
                accent: {
                    DEFAULT: "hsl(var(--accent) / <alpha-value>)",
                    foreground: "hsl(var(--accent-foreground) / <alpha-value>)",
                },
                popover: {
                    DEFAULT: "hsl(var(--popover) / <alpha-value>)",
                    foreground: "hsl(var(--popover-foreground) / <alpha-value>)",
                },
                card: {
                    DEFAULT: "hsl(var(--card) / <alpha-value>)",
                    foreground: "hsl(var(--card-foreground) / <alpha-value>)",
                },
            },
            borderRadius: {
                lg: "var(--radius)",
                md: "calc(var(--radius) - 2px)",
                sm: "calc(var(--radius) - 4px)",
            },
            keyframes: {
                "accordion-down": {
                    from: { height: "0" },
                    to: { height: "var(--radix-accordion-content-height)" },
                },
                "accordion-up": {
                    from: { height: "var(--radix-accordion-content-height)" },
                    to: { height: "0" },
                },
            },
            animation: {
                "accordion-down": "accordion-down 0.2s ease-out",
                "accordion-up": "accordion-up 0.2s ease-out",
            },
        },
    },
};
```

The plugin adds following base styles with values based on your configured theme.

```css
:root {
    --background: 0 0% 100%;
    --foreground: 222.2 47.4% 11.2%;

    --muted: 210 40% 96.1%;
    --muted-foreground: 215.4 16.3% 46.9%;

    --popover: 0 0% 100%;
    --popover-foreground: 222.2 47.4% 11.2%;

    --card: 0 0% 100%;
    --card-foreground: 222.2 47.4% 11.2%;

    --border: 214.3 31.8% 91.4%;
    --input: 214.3 31.8% 91.4%;

    --primary: 222.2 47.4% 11.2%;
    --primary-foreground: 210 40% 98%;

    --secondary: 210 40% 96.1%;
    --secondary-foreground: 222.2 47.4% 11.2%;

    --accent: 210 40% 96.1%;
    --accent-foreground: 222.2 47.4% 11.2%;

    --destructive: 0 100% 50%;
    --destructive-foreground: 210 40% 98%;

    --ring: 215 20.2% 65.1%;

    --radius: 0.5rem;
    --font-sans: Inter;
}

.dark {
    --background: 224 71% 4%;
    --foreground: 213 31% 91%;

    --muted: 223 47% 11%;
    --muted-foreground: 215.4 16.3% 56.9%;

    --popover: 224 71% 4%;
    --popover-foreground: 215 20.2% 65.1%;

    --card: 224 71% 4%;
    --card-foreground: 213 31% 91%;

    --border: 216 34% 17%;
    --input: 216 34% 17%;

    --primary: 210 40% 98%;
    --primary-foreground: 222.2 47.4% 1.2%;

    --secondary: 222.2 47.4% 11.2%;
    --secondary-foreground: 210 40% 98%;

    --accent: 216 34% 17%;
    --accent-foreground: 210 40% 98%;

    --destructive: 0 63% 31%;
    --destructive-foreground: 210 40% 98%;

    --ring: 216 34% 17%;

    --radius: 0.5rem;
}

* {
    @apply border-border;
}

body {
    @apply bg-background text-foreground;
    font-feature-settings: "rlig" 1, "calt" 1;
}
```


CHANGELOG.md
# tailwindcss-shadcn-ui

## 0.2.0

### Minor Changes

-   [#5](https://github.com/tobiasfoerg/tailwindcss-shadcn-ui/pull/5) [`4f362a1`](https://github.com/tobiasfoerg/tailwindcss-shadcn-ui/commit/4f362a124a7d97a8f99712d95d000a8c2f1cc847) Thanks [@tobiasfoerg](https://github.com/tobiasfoerg)! - Add fontFamily variable to plugin

## 0.1.0

### Minor Changes

-   [#2](https://github.com/tobiasfoerg/tailwindcss-shadcn-ui/pull/2) [`3a57061`](https://github.com/tobiasfoerg/tailwindcss-shadcn-ui/commit/3a57061c0d4151e0a8b5e209d8358b0330bf85ed) Thanks [@tobiasfoerg](https://github.com/tobiasfoerg)! - Add factory function to create tailwindcss preset


src/__tests__/run.ts
import path from "path";
import postcss from "postcss";
import tailwind, { Config } from "tailwindcss";
import { expect } from "vitest";
import tailwindShadcnUi from "..";

export let css = String.raw;
export let html = String.raw;
export let javascript = String.raw;

export function run(input: string, config: Config, plugin = tailwind) {
	let { currentTestName } = expect.getState();

	config.plugins ??= [];
	if (!config.plugins.includes(tailwindShadcnUi)) {
		config.plugins.push(tailwindShadcnUi);
	}

	return postcss(plugin(config)).process(input, {
		from: `${path.resolve(__filename)}?test=${currentTestName}`,
	});
}


src/__tests__/index.test.ts
import { describe, expect, test } from "vitest";
import { css, html, run } from "./run";

describe("plugin", () => {
	test("#1", async () => {
		const config = {
			content: [
				{
					raw: html``,
				},
			],
		};

		const input = css`
			@tailwind base;
		`;

		const result = await run(input, config);

		expect(result.css).toContain(`* {
  border-color: hsl(var(--border) / <alpha-value>);
}`);

		expect(result.css).toContain(`:root {
  --background: 0 0% 100%;
  --foreground: 222.2 47.4% 11.2%;
  --muted: 210 40% 96.1%;
  --muted-foreground: 215.4 16.3% 46.9%;
  --popover: 0 0% 100%;
  --popover-foreground: 222.2 47.4% 11.2%;
  --card: 0 0% 100%;
  --card-foreground: 222.2 47.4% 11.2%;
  --border: 214.3 31.8% 91.4%;
  --input: 214.3 31.8% 91.4%;
  --primary: 222.2 47.4% 11.2%;
  --primary-foreground: 210 40% 98%;
  --secondary: 210 40% 96.1%;
  --secondary-foreground: 222.2 47.4% 11.2%;
  --accent: 210 40% 96.1%;
  --accent-foreground: 222.2 47.4% 11.2%;
  --destructive: 0 100% 50%;
  --destructive-foreground: 210 40% 98%;
  --ring: 215 20.2% 65.1%;
  --radius: 0.5rem;
}`);
	});
});


src/__tests__/utils.test.ts
import { describe, expect, test } from "vitest";
import { defineTheme } from "..";

describe("defineConfig", () => {
	test("convert to css object", () => {
		const result = defineTheme({
			base: {
				radius: 20,
			},
			light: {
				background: [100, 0, 0],
				foreground: [0, 100, 0],
				muted: [0, 0, 100],
				mutedForeground: [100, 100, 0],
				popover: [0, 100, 100],
				popoverForeground: [100, 0, 100],
				card: [100, 100, 100],
				cardForeground: [0, 0, 0],
				border: [0, 0, 0],
				input: [0, 0, 0],
				primary: [0, 0, 0],
				primaryForeground: [0, 0, 0],
				secondary: [0, 0, 0],
				secondaryForeground: [0, 0, 0],
				accent: [0, 0, 0],
				accentForeground: [0, 0, 0],
				destructive: [0, 0, 0],
				destructiveForeground: [0, 0, 0],
				ring: [0, 0, 0],
			},
			dark: {
				background: "100 0% 0%",
				foreground: "0 100% 0%",
				muted: "0 0% 100%",
				mutedForeground: "100 100% 0%",
				popover: "0 100% 100%",
				popoverForeground: "100 0% 100%",
				card: "100 100% 100%",
				cardForeground: "100 0% 0%",
				border: "100 0% 0%",
				input: "100 0% 0%",
				primary: "100 0% 0%",
				primaryForeground: "100 0% 0%",
				secondary: "100 0% 0%",
				secondaryForeground: "100 0% 0%",
				accent: "100 0% 0%",
				accentForeground: "100 0% 0%",
				destructive: "100 0% 0%",
				destructiveForeground: "100 0% 0%",
				ring: "100 0% 0%",
			},
		});

		expect(result).toMatchObject({
			":root": {
				"--background": "100 0% 0%",
				"--foreground": "0 100% 0%",
				"--muted": "0 0% 100%",
				"--muted-foreground": "100 100% 0%",
				"--popover": "0 100% 100%",
				"--popover-foreground": "100 0% 100%",
				"--card": "100 100% 100%",
				"--card-foreground": "0 0% 0%",
				"--border": "0 0% 0%",
				"--input": "0 0% 0%",
				"--primary": "0 0% 0%",
				"--primary-foreground": "0 0% 0%",
				"--secondary": "0 0% 0%",
				"--secondary-foreground": "0 0% 0%",
				"--accent": "0 0% 0%",
				"--accent-foreground": "0 0% 0%",
				"--destructive": "0 0% 0%",
				"--destructive-foreground": "0 0% 0%",
				"--ring": "0 0% 0%",

				"--radius": "20px",
			},

			".dark": {
				"--background": "100 0% 0%",
				"--foreground": "0 100% 0%",
				"--muted": "0 0% 100%",
				"--muted-foreground": "100 100% 0%",
				"--popover": "0 100% 100%",
				"--popover-foreground": "100 0% 100%",
				"--card": "100 100% 100%",
				"--card-foreground": "100 0% 0%",
				"--border": "100 0% 0%",
				"--input": "100 0% 0%",
				"--primary": "100 0% 0%",
				"--primary-foreground": "100 0% 0%",
				"--secondary": "100 0% 0%",
				"--secondary-foreground": "100 0% 0%",
				"--accent": "100 0% 0%",
				"--accent-foreground": "100 0% 0%",
				"--destructive": "100 0% 0%",
				"--destructive-foreground": "100 0% 0%",
				"--ring": "100 0% 0%",
			},
		});
	});

	test("convert to css object light", () => {
		const result = defineTheme({
			base: {
				radius: "2rem",
			},
			light: {
				background: [100, 0, 0],
				foreground: [0, 100, 0],
				muted: [0, 0, 100],
				mutedForeground: [100, 100, 0],
				popover: [0, 100, 100],
				popoverForeground: [100, 0, 100],
				card: [100, 100, 100],
				cardForeground: [0, 0, 0],
				border: [0, 0, 0],
				input: [0, 0, 0],
				primary: [0, 0, 0],
				primaryForeground: [0, 0, 0],
				secondary: [0, 0, 0],
				secondaryForeground: [0, 0, 0],
				accent: [0, 0, 0],
				accentForeground: [0, 0, 0],
				destructive: [0, 0, 0],
				destructiveForeground: [0, 0, 0],
				ring: [0, 0, 0],
			},
		});

		expect(result).toMatchObject({
			":root": {
				"--background": "100 0% 0%",
				"--foreground": "0 100% 0%",
				"--muted": "0 0% 100%",
				"--muted-foreground": "100 100% 0%",
				"--popover": "0 100% 100%",
				"--popover-foreground": "100 0% 100%",
				"--card": "100 100% 100%",
				"--card-foreground": "0 0% 0%",
				"--border": "0 0% 0%",
				"--input": "0 0% 0%",
				"--primary": "0 0% 0%",
				"--primary-foreground": "0 0% 0%",
				"--secondary": "0 0% 0%",
				"--secondary-foreground": "0 0% 0%",
				"--accent": "0 0% 0%",
				"--accent-foreground": "0 0% 0%",
				"--destructive": "0 0% 0%",
				"--destructive-foreground": "0 0% 0%",
				"--ring": "0 0% 0%",

				"--radius": "2rem",
			},
		});
	});
});


src/index.ts
import { Config } from "tailwindcss";
// @ts-expect-error: no type definitions
import tailwindAnimate from "tailwindcss-animate";
import { fontFamily } from "tailwindcss/defaultTheme";
import plugin from "tailwindcss/plugin";
import { PluginOptions } from "./lib/types";
import { DEFAULT } from "./themes";

export { defineTheme } from "./lib/utils";
export type {
	ThemeColors as Colors,
	CssColor as HslValue,
	PluginOptions,
	Radius,
	Theme,
	ThemeConfig,
} from "./lib/types";

export function createPreset(options: PluginOptions = {}): Config {
	return {
		darkMode: ["class"],
		content: [],
		plugins: [tailwindAnimate, shadcnPlugin(options)],
	};
}

const shadcnPlugin = plugin.withOptions<PluginOptions>(
	function (options = {}) {
		options.theme ??= DEFAULT;

		return function ({ addBase }) {
			addBase({
				...options.theme,
			});

			addBase({
				"*": {
					"@apply border-border": {},
				},
				body: {
					"@apply bg-background text-foreground": {},
					"font-feature-settings": `"rlig" 1, "calt" 1`,
				},
			});
		};
	},
	function (_options) {
		return {
			theme: {
				container: {
					center: true,
					padding: "2rem",
					screens: {
						"2xl": "1400px",
					},
				},
				extend: {
					fontFamily: {
						sans: ["var(--font-sans)", ...fontFamily.sans],
					},
					colors: {
						border: "hsl(var(--border) / <alpha-value>)",
						input: "hsl(var(--input) / <alpha-value>)",
						ring: "hsl(var(--ring) / <alpha-value>)",
						background: "hsl(var(--background) / <alpha-value>)",
						foreground: "hsl(var(--foreground) / <alpha-value>)",
						primary: {
							DEFAULT: "hsl(var(--primary) / <alpha-value>)",
							foreground: "hsl(var(--primary-foreground) / <alpha-value>)",
						},
						secondary: {
							DEFAULT: "hsl(var(--secondary) / <alpha-value>)",
							foreground: "hsl(var(--secondary-foreground) / <alpha-value>)",
						},
						destructive: {
							DEFAULT: "hsl(var(--destructive) / <alpha-value>)",
							foreground: "hsl(var(--destructive-foreground) / <alpha-value>)",
						},
						muted: {
							DEFAULT: "hsl(var(--muted) / <alpha-value>)",
							foreground: "hsl(var(--muted-foreground) / <alpha-value>)",
						},
						accent: {
							DEFAULT: "hsl(var(--accent) / <alpha-value>)",
							foreground: "hsl(var(--accent-foreground) / <alpha-value>)",
						},
						popover: {
							DEFAULT: "hsl(var(--popover) / <alpha-value>)",
							foreground: "hsl(var(--popover-foreground) / <alpha-value>)",
						},
						card: {
							DEFAULT: "hsl(var(--card) / <alpha-value>)",
							foreground: "hsl(var(--card-foreground) / <alpha-value>)",
						},
					},
					borderRadius: {
						lg: "var(--radius)",
						md: "calc(var(--radius) - 2px)",
						sm: "calc(var(--radius) - 4px)",
					},
					keyframes: {
						"accordion-down": {
							from: { height: "0" },
							to: { height: "var(--radix-accordion-content-height)" },
						},
						"accordion-up": {
							from: { height: "var(--radix-accordion-content-height)" },
							to: { height: "0" },
						},
					},
					animation: {
						"accordion-down": "accordion-down 0.2s ease-out",
						"accordion-up": "accordion-up 0.2s ease-out",
					},
				},
			},
		};
	}
);

export default shadcnPlugin;


src/lib/types.ts
export type PluginOptions = {
	theme?: Theme;
};
export type Hue = number;
export type Saturation = number;
export type Lightness = number;
export type HexString = `#${string}`;
export type HslString = `${Hue} ${Saturation}% ${Lightness}%`;
export type HslTuple = [hue: Hue, saturation: Saturation, lightness: Lightness];
export type CssColor = HslString | HslTuple | HexString;
export type Radius = string | `${number}rem` | number;

export type ThemeColors = {
	/**
	 * Default backgrounds color of \<body />...etc
	 */
	"--background": CssColor;
	/**
	 * Default foregrounds color of \<body />...etc
	 */
	"--foreground": CssColor;

	/**
	 * Muted backgrounds such as \<TabsList />, \<Skeleton /> and \<Switch />
	 */
	"--muted": CssColor;
	/**
	 * Muted foregrounds such as \<TabsList />, \<Skeleton /> and \<Switch />
	 */
	"--muted-foreground": CssColor;

	"--popover": CssColor;
	"--popover-foreground": CssColor;

	"--card": CssColor;
	"--card-foreground": CssColor;

	"--border": CssColor;
	"--input": CssColor;

	"--primary": CssColor;
	"--primary-foreground": CssColor;

	"--secondary": CssColor;
	"--secondary-foreground": CssColor;

	"--accent": CssColor;
	"--accent-foreground": CssColor;

	"--destructive": CssColor;
	"--destructive-foreground": CssColor;

	"--ring": CssColor;
};

type ConvertKey<S extends string> = S extends `--${infer T}-${infer U}`
	? `${T}${Capitalize<ConvertKey<U>>}`
	: S extends `--${infer T}`
	? T
	: S;

export type ColorsToCamleCase<T> = T extends object
	? {
			[P in keyof T as ConvertKey<P & string>]: T[P];
	  }
	: T;

export type ThemeConfig = {
	base?: {
		radius?: Radius;
		fontFamily?: string | string[];
	};
	light: ColorsToCamleCase<ThemeColors>;
	dark: ColorsToCamleCase<ThemeColors>;
};

export type Theme = {
	/**
	 * Default theme
	 */
	":root": ToHslString<ThemeColors> & {
		"--radius": string;
		"--font-sans"?: string;
	};
	/**
	 * Dark theme
	 */
	".dark"?: ToHslString<ThemeColors>;
};

export type ToHslString<T> = {
	[P in keyof T]: HslString;
};


src/lib/utils.ts
import { SetOptional } from "type-fest";
import {
	ColorsToCamleCase,
	HexString,
	HslString,
	HslTuple,
	Theme,
	ThemeColors,
	ThemeConfig,
	ToHslString,
} from "./types";

export function defineTheme<T extends SetOptional<ThemeConfig, "dark">>(config: T): Theme {
	const { base = {}, dark, light } = config;
	return {
		":root": {
			"--radius": base.radius ? (typeof base.radius === "number" ? `${base.radius}px` : base.radius) : "0",
			...(base.fontFamily
				? { "--font-sans": Array.isArray(base.fontFamily) ? base.fontFamily.join(", ") : base.fontFamily }
				: {}),
			...toCSSRuleObject(light),
		},
		...(dark ? { ".dark": toCSSRuleObject(dark) } : {}),
	};
}

function toCSSRuleObject(colors: ColorsToCamleCase<ThemeColors>): ToHslString<ThemeColors> {
	return Object.entries(colors).reduce<ToHslString<ThemeColors>>((agg, [key, value]) => {
		let color: HslString;
		if (typeof value === "string") {
			if (isHexColor(value)) {
				color = hslToSting(rgbToHsl(value));
			} else if (isHslColor(value)) {
				color = value;
			} else {
				throw new Error(`Invalid color value: ${value}`);
			}
		} else {
			color = `${value[0]} ${value[1]}% ${value[2]}%`;
		}
		agg[("--" + key.replace(/[A-Z]/g, (letter) => `-${letter.toLowerCase()}`)) as keyof ToHslString<ThemeColors>] =
			color;
		return agg;
	}, {} as any);
}

function isHexColor(value: string): value is HexString {
	return /^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$/.test(value);
}

function isHslColor(value: string): value is HslString {
	return /^\d{1,3}\s\d{1,3}%\s\d{1,3}%\)$/.test(value);
}

function hslToSting(value: HslTuple): HslString {
	return `${value[0]} ${value[1]}% ${value[2]}%`;
}

function rgbToHsl(hex: HexString): HslTuple {
	const r = parseInt(hex.substring(1, 3), 16) / 255;
	const g = parseInt(hex.substring(3, 5), 16) / 255;
	const b = parseInt(hex.substring(5, 7), 16) / 255;
	const max = Math.max(r, g, b);
	const min = Math.min(r, g, b);
	let h = 0;
	let s = 0;
	let l = (max + min) / 2;

	if (max === min) {
		h = s = 0; // achromatic
	} else {
		var d = max - min;
		s = l > 0.5 ? d / (2 - max - min) : d / (max + min);
		switch (max) {
			case r:
				h = (g - b) / d + (g < b ? 6 : 0);
				break;
			case g:
				h = (b - r) / d + 2;
				break;
			case b:
				h = (r - g) / d + 4;
				break;
		}
		h /= 6;
	}

	return [h, s * 360, l * 360];
}


src/themes/default.ts
import { Theme } from "~/lib/types";

export const DEFAULT: Theme = {
	":root": {
		"--background": "0 0% 100%",
		"--foreground": "222.2 47.4% 11.2%",

		"--muted": "210 40% 96.1%",
		"--muted-foreground": "215.4 16.3% 46.9%",

		"--popover": "0 0% 100%",
		"--popover-foreground": "222.2 47.4% 11.2%",

		"--card": "0 0% 100%",
		"--card-foreground": "222.2 47.4% 11.2%",

		"--border": "214.3 31.8% 91.4%",
		"--input": "214.3 31.8% 91.4%",

		"--primary": "222.2 47.4% 11.2%",
		"--primary-foreground": "210 40% 98%",

		"--secondary": "210 40% 96.1%",
		"--secondary-foreground": "222.2 47.4% 11.2%",

		"--accent": "210 40% 96.1%",
		"--accent-foreground": "222.2 47.4% 11.2%",

		"--destructive": "0 100% 50%",
		"--destructive-foreground": "210 40% 98%",

		"--ring": "215 20.2% 65.1%",

		"--radius": "0.5rem",
	},

	".dark": {
		"--background": "0 0% 100%",
		"--foreground": "222.2 47.4% 11.2%",

		"--muted": "210 40% 96.1%",
		"--muted-foreground": "215.4 16.3% 46.9%",

		"--popover": "0 0% 100%",
		"--popover-foreground": "222.2 47.4% 11.2%",

		"--card": "0 0% 100%",
		"--card-foreground": "222.2 47.4% 11.2%",

		"--border": "214.3 31.8% 91.4%",
		"--input": "214.3 31.8% 91.4%",

		"--primary": "222.2 47.4% 11.2%",
		"--primary-foreground": "210 40% 98%",

		"--secondary": "210 40% 96.1%",
		"--secondary-foreground": "222.2 47.4% 11.2%",

		"--accent": "210 40% 96.1%",
		"--accent-foreground": "222.2 47.4% 11.2%",

		"--destructive": "0 100% 50%",
		"--destructive-foreground": "210 40% 98%",

		"--ring": "215 20.2% 65.1%",
	},
};
